google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/SpannerRpcMetricsTest.java

```diff
@@ -0,0 +1,304 @@
+    spannerNoHeader =
+        createSpannerOptions(addressNoHeader, serverNoHeader, inMemoryMetricReader).getService();
+    databaseClientNoHeader =
+        spannerNoHeader.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+  }
+
+  @AfterClass
+  public static void stopServer() throws InterruptedException {
+    if (spanner != null) {
+      spanner.close();
+      server.shutdown();
+      server.awaitTermination();
+    }
+
+    if (spannerNoHeader != null) {
+      spannerNoHeader.close();
+      serverNoHeader.shutdown();
+      serverNoHeader.awaitTermination();
+    }
+  }
+
+  @After
+  public void reset() {
+    mockSpanner.reset();
+    mockSpannerNoHeader.reset();
+  }
+
+  @Test
+  public void testGfeLatencyExecuteSql() throws InterruptedException {
+    databaseClient
+        .readWriteTransaction()
+        .run(transaction -> transaction.executeUpdate(UPDATE_FOO_STATEMENT));
+
+    ;
+    double latency =
+        getGfeLatencyMetric(getMetricData("gfe_latency"), "google.spanner.v1.Spanner/ExecuteSql");
+    assertEquals(fakeServerTiming.get(), latency, 0);
+  }
+
+  @Test
+  public void testGfeMissingHeaderExecuteSql() throws InterruptedException {
+    databaseClient
+        .readWriteTransaction()
+        .run(transaction -> transaction.executeUpdate(UPDATE_FOO_STATEMENT));
+    long count =
+        getHeaderLatencyMetric(
+            getMetricData("gfe_header_missing_count"), "google.spanner.v1.Spanner/ExecuteSql");
+    assertEquals(0, count);
+
+    databaseClientNoHeader
+        .readWriteTransaction()
+        .run(transaction -> transaction.executeUpdate(UPDATE_FOO_STATEMENT));
+    long count1 =
+        getHeaderLatencyMetric(
+            getMetricData("gfe_header_missing_count"), "google.spanner.v1.Spanner/ExecuteSql");
+    assertEquals(1, count1);
+  }
+
+  private static SpannerOptions createSpannerOptions(
+      InetSocketAddress address, Server server, InMemoryMetricReader inMemoryMetricReader) {
+    SdkMeterProvider sdkMeterProvider =
+        SdkMeterProvider.builder().registerMetricReader(inMemoryMetricReader).build();
+    OpenTelemetry openTelemetry =
+        OpenTelemetrySdk.builder().setMeterProvider(sdkMeterProvider).build();
+    String endpoint = address.getHostString() + ":" + server.getPort();
+    return SpannerOptions.newBuilder()
+        .setProjectId("[PROJECT]")
+        // Set a custom channel configurator to allow http instead of https.
+        .setChannelConfigurator(
+            input -> {
+              input.usePlaintext();
+              return input;
+            })
+        .setHost("http://" + endpoint)
+        // Set static credentials that will return the static OAuth test token.
+        .setCredentials(STATIC_CREDENTIALS)
+        // Also set a CallCredentialsProvider. These credentials should take precedence above
+        // the static credentials.
+        .setCallCredentialsProvider(() -> MoreCallCredentials.from(VARIABLE_CREDENTIALS))
```

Comment:
I don't think you need any of these. Instead, you can just call `setCredentials(NoCredentials.getInstance())`.

Reply:
Done